lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivationSpec.hs

```diff
@@ -0,0 +1,315 @@
+                        , "track", "either", "maid", "organ", "sock"
+                        ]
+            res `shouldSatisfy` isRight
+
+        it "successfully parse 15 words in [9,12,15]" $ do
+            let res = fromMnemonic @'[9,12,15] @"testing"
+                        ["cushion", "anxiety", "oval", "village", "choose"
+                        , "shoot", "over", "behave", "category", "cruise"
+                        , "track", "either", "maid", "organ", "sock"
+                        ]
+            res `shouldSatisfy` isRight
+
+    describe "Keys storing and retrieving roundtrips" $ do
+        it "XPriv" (property prop_roundtripXPriv)
+        it "XPub" (property prop_roundtripXPub)
+
+
+{-------------------------------------------------------------------------------
+                               Properties
+-------------------------------------------------------------------------------}
+
+
+prop_succMaxBoundHardIx :: Property
+prop_succMaxBoundHardIx = expectFailure $
+    property $ succ (maxBound @(Index 'Hardened _)) `seq` ()
+
+prop_predMinBoundHardIx :: Property
+prop_predMinBoundHardIx = expectFailure $
+    property $ pred (minBound @(Index 'Hardened _)) `seq` ()
+
+prop_succMaxBoundSoftIx :: Property
+prop_succMaxBoundSoftIx = expectFailure $
+    property $ succ (maxBound @(Index 'Soft _)) `seq` ()
+
+prop_predMinBoundSoftIx :: Property
+prop_predMinBoundSoftIx = expectFailure $
+    property $ pred (minBound @(Index 'Soft _)) `seq` ()
+
+prop_roundtripEnumIndexHard :: Index 'Hardened 'AccountK -> Property
+prop_roundtripEnumIndexHard ix =
+    (toEnum . fromEnum) ix === ix .&&. (toEnum . fromEnum . getIndex) ix === ix
+
+prop_roundtripEnumIndexSoft :: Index 'Soft 'AddressK -> Property
+prop_roundtripEnumIndexSoft ix =
+    (toEnum . fromEnum) ix === ix .&&. (toEnum . fromEnum . getIndex) ix === ix
+
+prop_roundtripXPriv
+    :: (Key 'RootK XPrv, Hash "encryption")
+    -> Property
+prop_roundtripXPriv xpriv = do
+    let xpriv' = (deserializeXPrv . serializeXPrv) xpriv
+    xpriv' === Right xpriv
+
+prop_roundtripXPub
+    :: Key 'RootK XPrv
+    -> Property
+prop_roundtripXPub xpriv = do
+    let xpub = publicKey xpriv
+    let xpub' = (deserializeXPub . serializeXPub) xpub
+    xpub' === Right xpub
+
+prop_passphraseRoundtrip
+    :: Passphrase "encryption"
+    -> Property
+prop_passphraseRoundtrip pwd = monadicIO $ liftIO $ do
+    hpwd <- encryptPassphrase pwd
+    checkPassphrase pwd hpwd `shouldBe` Right ()
+
+prop_passphraseRoundtripFail
+    :: (Passphrase "encryption", Passphrase "encryption")
+    -> Property
+prop_passphraseRoundtripFail (p, p') =
+    p /= p' ==> monadicIO $ liftIO $ do
+        hp <- encryptPassphrase p
+        checkPassphrase p' hp `shouldBe` Left ErrWrongPassphrase
+
+prop_passphraseHashMalformed
+    :: Passphrase "encryption"
+    -> Property
+prop_passphraseHashMalformed pwd = monadicIO $ liftIO $ do
+    checkPassphrase pwd (Hash mempty) `shouldBe` Left ErrWrongPassphrase
+
+
```

Comment:
and here!

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

Reply:
Thanks, fixed.